File: js/Chess/PieceType.js
import Location from './Location'

export default class PieceType {
    static Rook = new PieceType(true, 'R', new Location(1, 0), new Location(0, 1));
    static King = new PieceType(false, 'K', new Location(1, 1), new Location(1, 0), new Location(0, 1));
    static Queen = new PieceType(true, 'Q', new Location(1, 1), new Location(1, 0), new Location(0, 1));
    static Bishop = new PieceType(true, 'B', new Location(1, 1));
    static Knight = new PieceType(false, 'N', new Location(2, 1), new Location(1, 2));

    constructor(infinite, sign, ...directions) {
        this.infinite = infinite;
        this.sign = sign;
        this.directions = getPermutations(directions);
    }

    getMoves(from, board) {
        return this.directions.reduce((res, direction) => res
            .concat(this.movesInOneDirection(from, board, direction)),
        []);
    }

    movesInOneDirection(from, board, direction) {
        const piece = board.getPiece(from);
        const result = [];
        let distance = 1;

        do {
            const to = sum(from, multiply(direction, distance));
            if (!board.contains(to)) {
                break;
            }
            const destinationPiece = board.getPiece(to);
            if (destinationPiece === null) {
                result.push(to);
            } else {
                if (destinationPiece.color !== piece.color) {
                    result.push(to);
                }
                break;
            }
        } while (this.infinite && distance++);

        return result;
    }

    toString() {
        return this.sign;
    }
}

function equals(a, b) {
    return a.x === b.x && a.y === b.y;
}

function multiply(a, b) {
    if (typeof b === 'number') {
        return new Location(a.x * b, a.y * b);
    }
    return new Location(a.x * b.x, a.y * b.y);
}

function sum(a, b) {
    return new Location(a.x + b.x, a.y + b.y);
}

function contains(arr, val) {
    return arr.some(item => equals(val, item));
}

function getPermutations(directions) {
    const deltas = [
        {x: 1, y: 1},
        {x: -1, y: 1},
        {x: 1, y: -1},
        {x: -1, y: -1},
    ];

    const permutations = [];
    for (const direction of directions) {
        for (const delta of deltas) {
            const current = multiply(direction, delta);
            if (!contains(permutations, current)) {
                permutations.push(current);
            }
        }
    }
    return permutations;
}

File: python/Chess/piece_type.py
from __future__ import annotations
from typing import List
from location import Location
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from board import Board


class PieceType:
    def __init__(self, infinite: bool, sign: str, directions: List[Location]):
        self.infinite = infinite
        self.sign = sign
        self.directions = get_permutations(directions)

    def get_movies(self, location_from: Location, board: Board):
        result = []
        for direction in self.directions:
            result.extend(self.moves_in_one_direction(location_from, board, direction))
        return result

    def moves_in_one_direction(
        self, location_from: Location, board: Board, direction: Location
    ):
        piece = board.get_piece(location_from)
        result = []
        distance = 1

        while distance:
            to = location_from + direction * distance
            if not board.contains(to):
                break
            destination_piece = board.get_piece(to)
            if destination_piece is None:
                result.append(to)
            else:
                if destination_piece.color != piece.color:
                    result.append(to)
                break
            distance = distance + 1 if self.infinite else 0
        return result

    def __str__(self):
        return self.sign


def get_permutations(directions: List[Location]) -> List[Location]:
    deltas = [
        Location(1, 1),
        Location(1, -1),
        Location(-1, 1),
        Location(-1, -1),
    ]

    permutations = []
    for direction in directions:
        for delta in deltas:
            current = direction * delta
            if current not in permutations:
                permutations.append(current)

    return permutations


Rook = PieceType(True, 'R', [Location(1, 0), Location(0, 1)])
King = PieceType(False, 'K', [Location(1, 1),  Location(1, 0),  Location(0, 1)])
Queen = PieceType(True, 'Q', [Location(1, 1),  Location(1, 0),  Location(0, 1)])
Bishop = PieceType(True, 'B', [Location(1, 1)])
Knight = PieceType(False, 'N', [Location(2, 1),  Location(1, 2)])
